### PR TITLE
[#12382] feat(lance): Support AddColumn via Gravitino API

### LINKED ISSUE #12382: [FEATURE] Support adding nullable columns to Lance tables via Gravitino API
### Describe the feature

Support `TableChange.AddColumn` for Lance tables in the Generic Lakehouse Catalog through the Gravitino table API and Java client.

Multiple `AddColumn` changes in one request should be applied in a single Lance schema commit. Existing rows should expose the newly added nullable columns as `NULL`.

### Motivation

Schema evolution is a common requirement for Lance tables. Currently, users cannot add Lance columns through Gravitino and must modify the dataset through an external Lance client, which may cause the Lance schema and Gravitino metadata to become inconsistent.

Providing this capability through Gravitino keeps schema changes and table metadata under the same catalog operation.

### Describe the solution

- Validate all changes before modifying the Lance dataset.
- Initially support only top-level, nullable columns appended to the end of the schema.
- Reject default values, auto-increment columns, custom positions, and nested columns.
- Reject requests that combine `AddColumn` with other table-change types.
- Convert all new columns to Arrow fields and call `Dataset.addColumns` once.
- Verify the Lance schema before and after the operation.
- Persist the new columns and Lance dataset version in Gravitino with optimistic-lock protection.
- Attempt to remove the added Lance columns if result validation or Gravitino metadata persistence fails.
- Add unit and integration tests covering batching, conflicts, failure recovery, and `NULL` backfill.
- Document the supported behavior and limitations.

### Additional context

The first phase does not include:

- Nested columns
- Non-nullable columns
- Default-value backfill
- `FIRST` or `AFTER` positions
- Auto-increment columns
- A Lance REST `/add_columns` endpoint

Lance and the Gravitino entity store do not share a distributed transaction, so failure compensation is best-effort.

### PR DESCRIPTION
**Pull request by bbiiaaoo**
### What changes were proposed in this pull request?

This PR adds first-phase `AddColumn` support for Lance tables through the Gravitino table API.

The implementation:

- Supports nullable, top-level, append-only columns without default values or auto-increment.
- Batches multiple `AddColumn` changes into one `Dataset.addColumns` call.
- Uses Lance's native `NULL` backfill for existing rows.
- Hydrates declared or empty stored metadata before applying the incremental column changes.
- Reuses `super.alterTable` to persist the requested columns, `lance.version`, and remove `lance.declared`.

The implementation follows the existing Lance alter-table consistency model. It does not introduce custom metadata CAS, physical rollback, or strict schema reconstruction.

### Why are the changes needed?

Lance tables currently cannot add columns through the Gravitino table API.

This change provides initial AddColumn support while keeping Lance as the source of truth and keeping broader Lance-to-Gravitino schema reconciliation outside the scope of this PR.

Fix: #12382

### Does this PR introduce _any_ user-facing change?

Yes.

Users can add nullable, top-level columns to Lance tables through the Gravitino table API. Existing rows are backfilled with `NULL`.

The Lance REST `/add_columns` endpoint is not included in this change.

### How was this patch tested?

- Added unit tests for batched AddColumn, validation, declared/empty metadata hydration, and zero-column declared tables.
- Added integration coverage verifying that historical rows are backfilled with `NULL` and multiple columns produce one Lance version.
- Added integration coverage for direct AddColumn on a declared table without a preceding `loadTable`.

Commands:

- `./gradlew spotlessApply`
- `./gradlew :catalogs:catalog-lakehouse-generic:test --tests org.apache.gravitino.catalog.lakehouse.lance.TestLanceTableOperations -PskipITs`
- `./gradlew :catalogs:catalog-lakehouse-generic:check -PskipITs`

### CODE REVIEW

**Comment by bbiiaaoo** [comment]
Hi @yuqi1129 @FANNG1 , could you please take a look at this PR when you have time?

**Comment by FANNG1** [comment]
### Code review

Found 1 issue:

1. `AddColumn` deterministically fails on declared-only or empty-stored-columns tables. `alterTable` loads the entity via `loadTableEntity(ident)`, a raw `store.get()` that bypasses the repair-on-load logic in `loadTable`, and `TableOperationDispatcher.alterTable` does not call `loadTable` beforehand. The new `addColumns` then builds `expectedCurrentFields` from `table.columns()` and hard-checks it against the real dataset schema, so for a declared table (`lance.declared=true`, created with `new Column[0]`) or a table whose stored columns are still empty — both normal states under the default `DECLARED_AND_EMPTY` refresh mode, per the class Javadoc — the pre-check throws a misleading `OptimisticLockException` about schema inconsistency. The pre-existing DeleteColumn/RenameColumn/AddIndex paths are unaffected because they never consult `table.columns()`. No test covers add-column against a declared or empty-stored-columns table.

https://github.com/apache/gravitino/blob/dccd7db3387b233f62a6b41370445e5adad502e5/catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java#L309-L313

https://github.com/apache/gravitino/blob/dccd7db3387b233f62a6b41370445e5adad502e5/catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java#L831-L839

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by bbiiaaoo** [comment]
@FANNG1 Thanks for catching this. Fixed by hydrating the Lance schema before handling AddColumn and reloading the TableEntity afterward as the optimistic-lock snapshot. Added regression tests for both declared tables and non-declared tables with empty stored columns.

**Comment by bbiiaaoo** [comment]
Hi @FANNG1  @yuqi1129 , would it be acceptable to simplify the AddColumn implementation by following the existing Lance alter-table flow?

The proposed flow is:

1. Load the table and hydrate the Lance schema if necessary.
2. Add columns through `Dataset.addColumns`.
3. Reuse `super.alterTable` to update the Gravitino columns and `lance.version`.

This would remove the custom metadata CAS and physical rollback logic, and would provide the same best-effort consistency model as the existing DeleteColumn/RenameColumn paths.

Would this simpler approach be acceptable for the first-phase AddColumn support?

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 1]
```diff
@@ -301,9 +307,29 @@ public Table createTable(
...
+    TableEntity loadedEntity = loadTableEntity(ident);
+    Table loadedTable = toGenericTable(loadedEntity);
+    validateFieldsToAdd(loadedTable, fieldsToAdd);
+    long version = handleLanceTableChange(loadedTable, changes, fieldsToAdd);
+
+    if (!fieldsToAdd.isEmpty()) {
+      try {
+        return persistAddedColumns(ident, loadedEntity, changes, version);
+      } catch (RuntimeException metadataFailure) {
+        rollbackAddedColumns(loadedTable, fieldsToAdd, metadataFailure);
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: Would you roll back the column created in the Lance dataset when writing column information fails?

**Comment by yuqi1129** [comment]
> Hi @FANNG1 @yuqi1129 , would it be acceptable to simplify the AddColumn implementation by following the existing Lance alter-table flow?
> 
> The proposed flow is:
> 
> 1. Load the table and hydrate the Lance schema if necessary.
> 2. Add columns through `Dataset.addColumns`.
> 3. Reuse `super.alterTable` to update the Gravitino columns and `lance.version`.
> 
> This would remove the custom metadata CAS and physical rollback logic, and would provide the same best-effort consistency model as the existing DeleteColumn/RenameColumn paths.
> 
> Would this simpler approach be acceptable for the first-phase AddColumn support?

It's fine currently.  In fact, we need to review the Lance directory as the single source of truth; no matter how much effort we put in, there is still a high possibility that the data in Gravitino and Lance differ.

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 2]
```diff
@@ -797,6 +836,35 @@ long handleLanceTableChange(Table table, TableChange[] changes) {
     }
   }
 
+  private long addColumns(
+      Table table, List<Field> fieldsToAdd, String location, Map<String, String> storageOptions) {
+    List<Field> expectedCurrentFields = convertColumnsToArrowSchema(table.columns()).getFields();
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: The "expected" schema is rebuilt from Gravitino metadata through `LanceDataTypeConverter` and then compared to the live Lance schema with an exact positional field comparison (see `fieldMatches` below). But `toGravitino` -> `fromGravitino` is not round-trip lossless, so several classes of perfectly consistent tables will fail the pre-check at line 847 and never be able to add a column again:

- **Timestamp with time zone**: `toGravitino` maps any tz to `TimestampType.withTimeZone(p)`, and `fromGravitino` always re-emits `Timestamp(unit, "UTC")` (the converter even carries `// todo: need timeZoneId for timestamp with time zone`). A dataset with `timestamp[us, tz=Asia/Shanghai]` never matches.
- **List**: `toGravitino` drops the list child field name, and `toArrowField` hard-codes it back as `"element"`, while datasets written by pyarrow / arrow-rs name it `"item"`.
- **Decimal**: `fromGravitino` always emits bit width 128, so a `decimal256` column never matches.

The resulting `OptimisticLockException` surfaces as HTTP 409, which tells the client to retry, but the condition is permanent — under the default `DECLARED_AND_EMPTY` refresh mode a non-empty table is never re-hydrated, so there's no supported way to clear it.

The same strict comparison is reused for the post-add check at line 851, which has the mirror problem: if Lance normalizes the written field in any way, an add that actually succeeded gets rolled back and reported as a failure.

Would it be possible to compare against the schema actually read from the dataset (e.g. snapshot it before the add and diff that against the post-add schema), rather than against a re-derived one?

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java (hunk 1 above)
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: This rollback can leave Gravitino metadata listing columns that no longer exist physically.

`persistAddedColumns` fails the CAS whenever `!current.equals(expectedEntity)`. With `lance.schema-refresh-mode=VERSION_CHECK`, a concurrent `loadTable` landing in the window between `dataset.addColumns()` committing and the CAS running will hydrate the *new* schema into the entity store and bump `lance.version` via `repairTableMetadata`. That both (a) makes this CAS fail and (b) means the store already contains the added columns. We then call `rollbackAddedColumns`, which physically drops them from the dataset, and rethrow — net result is metadata referencing columns Lance doesn't have, and under the default `DECLARED_AND_EMPTY` mode a non-empty table is never re-checked, so it stays that way.

The dispatcher only takes a READ tree-lock for non-rename changes (`TableOperationDispatcher.alterTable`), so this interleaving isn't excluded.

One option: make the CAS compare only the fields that actually matter (columns + `lance.version`) and skip the physical rollback when the stored schema already reflects the add.

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 3]
```diff
@@ -811,6 +879,257 @@ Dataset openDataset(String location, Map<String, String> storageOptions) {
...
+    }
+    for (int i = 0; i < expectedFields.size(); i++) {
+      if (!fieldMatches(expectedFields.get(i), actualFields.get(i))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean fieldMatches(Field expected, Field actual) {
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: Minor / lower severity, but it lands on the same path: column types Lance/Arrow can't map back to a Gravitino type — notably `fixed_size_list` vector columns, arguably the flagship Lance type — are stored as `Types.ExternalType` whose `catalogString` is the serialized Arrow `Field`, *including its original name*.

`RenameColumn` is supported for Lance tables and rewrites only the `ColumnEntity` name, leaving the embedded JSON name stale. `convertColumnsToArrowSchema` at line 841 then hits the EXTERNAL branch of `toArrowField`, which does `Preconditions.checkArgument(name.equals(field.getName()), "expected field name %s but got %s")`. So after renaming a vector column, every subsequent `AddColumn` on that table dies with a confusing `IllegalArgumentException` before any Lance work happens.

Either rewrite the embedded name on rename, or override the name when reconstructing the Arrow field from `ExternalType`.

**Review by yuqi1129 — COMMENTED** [review]
Thanks for the work here! The add/verify/rollback sequence itself looks correct (I checked that `Dataset.checkoutLatest()` is in-place and `getSchema()` is not cached in lance-core 6.0.0). I left three comments, two of which I think can permanently break `AddColumn` on otherwise-healthy tables.

Also verified as fine, in case it saves anyone else the trip: no double rollback on the `persistAddedColumns` path, `ColumnPosition.defaultPos()` is a singleton so the `equals` position check holds over REST, `appendAddedColumns` position assignment is safe because `ManagedTableOperations.applyChanges` re-indexes, and `OptimisticLockException` maps to 409 rather than 500.

**Comment by FANNG1** [comment]
+1 to reusing the existing Lance `alterTable` flow for the first-phase `AddColumn` support.

One architectural concern remains: `super.alterTable` applies the requested changes incrementally to the Gravitino metadata, so it may still produce a Gravitino schema that differs from the schema actually committed by Lance, especially when the two sides have already drifted or Lance normalizes the resulting schema. Since the Lance dataset is the source of truth, replacing or reconciling the Gravitino schema from the actual underlying dataset after the physical change may be more robust.

I think that broader reconciliation change can be handled in a separate PR to keep this one focused. What do you think?

**Comment by bbiiaaoo** [comment]
Thanks @FANNG1 and @yuqi1129 for the suggestions.

I have simplified and force-pushed the implementation based on the latest `main`.

The updated implementation now:

1. Validates nullable, top-level, append-only `AddColumn` changes.
2. Hydrates declared or empty stored metadata before adding columns.
3. Batches the new fields into one `Dataset.addColumns` call.
4. Reuses `super.alterTable` to persist the requested columns and `lance.version`, and removes `lance.declared`.

I removed the strict Gravitino-to-Arrow schema reconstruction and comparison, custom metadata CAS, manual `TableEntity` construction, and physical rollback logic.

I agree that broader reconciliation from the Lance dataset back to Gravitino metadata should be handled in a separate PR, with Lance treated as the source of truth.

Unit tests, the relevant integration tests, and the module check pass locally. Could you please take another look?

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 4]
```diff
@@ -811,6 +837,78 @@ Dataset openDataset(String location, Map<String, String> storageOptions) {
...
+      return List.of();
+    }
+
+    Preconditions.checkArgument(
+        addColumnCount == changes.length,
+        "Lance AddColumn cannot be combined with other table changes");
+
+    Set<String> columnNames = new HashSet<>();
+    List<Field> fieldsToAdd = new ArrayList<>(changes.length);
+    for (TableChange change : changes) {
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: Is it possible to merge the loop above and the loop here?

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 5]
```diff
@@ -301,15 +304,24 @@ public Table createTable(
...
     // After making changes to the Lance dataset, we need to update the table metadata in
     // Gravitino. If there's any failure during this process, the code will throw an exception
     // and the update won't be applied in Gravitino.
-    TableChange[] metadataChanges = Arrays.copyOf(changes, changes.length + 1);
+    int internalChangeCount = fieldsToAdd.isEmpty() ? 1 : 2;
+    TableChange[] metadataChanges = Arrays.copyOf(changes, changes.length + internalChangeCount);
     metadataChanges[changes.length] =
         TableChange.setProperty(LanceConstants.LANCE_TABLE_VERSION, String.valueOf(version));
+    if (!fieldsToAdd.isEmpty()) {
+      metadataChanges[changes.length + 1] =
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: You can use `ArrayUtils.add()` directly with explicitly copying it.

Comment (yuqi1129) [reply catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: Another problem is why the index is `changes.length + 1`, not `changes.length` here?

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 6]
```diff
@@ -754,10 +766,24 @@ private GenericColumn toGenericColumn(ColumnEntity columnEntity) {
...
+  private long handleLanceTableChange(Table table, TableChange[] changes, List<Field> fieldsToAdd) {
     String location = table.properties().get(Table.PROPERTY_LOCATION);
     Map<String, String> storageOptions =
         LancePropertiesUtils.resolveLanceStorageOptions(catalogProperties, table.properties());
     try (Dataset dataset = openDataset(location, storageOptions)) {
+      if (!fieldsToAdd.isEmpty()) {
+        // Adding all fields in one call creates one Lance schema version and backfills existing
+        // rows with null for the new nullable columns.
+        dataset.addColumns(fieldsToAdd);
+        dataset.checkoutLatest();
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java:800]: How do you handle the case where there exist `addColumn` and `dropColumn` or `AddIndex` at the same time? Have you already excluded such scenarios?

File: catalogs/catalog-lakehouse-generic/src/test/java/org/apache/gravitino/catalog/lakehouse/lance/TestLanceTableOperations.java [hunk 7]
```diff
@@ -473,6 +475,168 @@ public void testAlterTablePersistsUpdatedLanceVersion() throws Exception {
     Assertions.assertEquals("9", storedTable.get().properties().get(LANCE_TABLE_VERSION));
   }
 
+  @Test
+  public void testAlterTableAddsNullableColumnsInSingleLanceCommit() throws Exception {
```
Comment (yuqi1129) [inline catalogs/catalog-lakehouse-generic/src/test/java/org/apache/gravitino/catalog/lakehouse/lance/TestLanceTableOperations.java:640]: Have you covered the case where we changed Lance to succeed but failed to write metadata to Gravitino, and then It's still okay when we load the table?

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java (hunk 4 above)
Comment (bbiiaaoo) [reply catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: Thanks for the suggestion. I merged AddColumn detection, validation, and Arrow field construction into a single loop.

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java (hunk 5 above)
Comment (bbiiaaoo) [reply catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: Done. I replaced the manual array copying and indexing with sequential ArrayUtils.add calls.

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java (hunk 6 above)
Comment (bbiiaaoo) [reply catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java:800]: Mixed AddColumn and other table changes are rejected because the Lance physical schema update must be completed separately from the existing alter-table operations.

I added explicit tests for AddColumn combined with DeleteColumn in both orders, as well as AddColumn combined with AddIndex.

File: catalogs/catalog-lakehouse-generic/src/test/java/org/apache/gravitino/catalog/lakehouse/lance/TestLanceTableOperations.java (hunk 7 above)
Comment (bbiiaaoo) [reply catalogs/catalog-lakehouse-generic/src/test/java/org/apache/gravitino/catalog/lakehouse/lance/TestLanceTableOperations.java:640]: I added a regression test for this failure window.

If the Lance add-columns commit succeeds but the Gravitino metadata update fails, a subsequent load in VERSION_CHECK mode detects the changed Lance version and refreshes both the schema and version from the underlying dataset.

Under the default DECLARED_AND_EMPTY mode, an existing non-empty stored schema is not automatically reconciled. Broader reconciliation in that mode can be addressed separately as part of the Lance-as-source-of-truth follow-up work.

**Comment by bbiiaaoo** [comment]
Hi @yuqi1129 , I have rebased this PR onto the latest main and resolved the conflict introduced by the recent Lance schema-hydration changes. The branch is mergeable again.

The focused unit tests and module checks pass locally. The workflows for the latest push also appear to be awaiting maintainer approval.

Could you please take another look and approve the workflows when you have time? Thanks!

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 8]
```diff
@@ -812,10 +826,24 @@ private GenericColumn toGenericColumn(ColumnEntity columnEntity) {
...
+        return dataset.getVersion().getId();
+      }
+
       for (TableChange change : changes) {
         if (change instanceof TableChange.DeleteColumn deleteColumn) {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: could you add `add column` here? move all add column related logic in one place

File: catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java [hunk 9]
```diff
@@ -812,10 +826,24 @@ private GenericColumn toGenericColumn(ColumnEntity columnEntity) {
    * @return the new version id of the Lance dataset after applying the changes
    */
   long handleLanceTableChange(Table table, TableChange[] changes) {
+    List<Field> fieldsToAdd = prepareFieldsToAdd(changes);
+    validateFieldsToAdd(table, fieldsToAdd);
+    return handleLanceTableChange(table, changes, fieldsToAdd);
+  }
+
+  private long handleLanceTableChange(Table table, TableChange[] changes, List<Field> fieldsToAdd) {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-generic/src/main/java/org/apache/gravitino/catalog/lakehouse/lance/LanceTableOperations.java]: `fieldsToAdd` seems not neccessary?

**Comment by FANNG1** [comment]
Please keeping this PR focus on the add column logic. Let's not try to solve Gravitino/Lance schema synchronization here, leave  #13339 and #13340 to handle these problems.  @yuqi1129 what do you think?

**Comment by yuqi1129** [comment]
> Please keeping this PR focus on the add column logic. Let's not try to solve Gravitino/Lance schema synchronization here, leave #13339 and #13340 to handle these problems. @yuqi1129 what do you think?

I think so. Schema synchronization is complicated, and we need to come out with a thorough solution.

**Comment by bbiiaaoo** [comment]
@FANNG1 @yuqi1129 Thanks, understood. I’ll narrow this PR to the AddColumn operation only. I’ll remove the declared/empty-schema hydration and metadata recovery changes, handle AddColumn in the existing table-change loop, and leave schema/version reconciliation and load-path semantics to #13339 and #13340.

**Comment by FANNG1** [comment]
LGTM, @yuqi1129 PTAL

**Review by FANNG1 — APPROVED** [review]

**Review by yuqi1129 — APPROVED** [review]
